### ISSUE TITLE
error: function lower(enum_users_status) does not exist

### ISSUE THREAD

**Issue by sgress454** [balderdashy/sails#4009]
_From @slavafomin on August 8, 2016 22:36_

Hello!

Thank you for this great library!

However, I'm having an error: `function lower(enum_users_status) does not exist`.

I'm not sure if this issue relates to this module or to waterline.

Here's my other comment (left unnoticed):
https://github.com/balderdashy/sails-postgresql/issues/99#issuecomment-236730862

---

The `enum_users_status` is a custom enumeration type used in our PostgreSQL database. The schema for it is:

``` pgsql
CREATE TYPE "enum_users_status" AS ENUM (
  'foo',
  'bar',
  'baz'
);
```

Such database type should be easily mapped to `string` or `text` type.

It's defined in my model as:

```
    status: {
      type: 'string',
      size: 20,
      enum: ['foo', 'bar', 'baz'],
      defaultsTo: 'foo',
      required: true
    }
```

What is the reason for this error? Such configuration works perfectly fine with Sequelize in our other project.

```
$ npm ls waterline
app
├─┬ sails@0.12.3
│ └── waterline@0.11.2 
└─┬ sails-hook-validation@0.4.6
  └── waterline@0.10.31
```

Please advise, thank you!

_Copied from original issue: balderdashy/sails-postgresql#256_

**Comment by sgress454** [comment]
_From @sailsbot on August 8, 2016 22:36_

@slavafomin Thanks for posting, we'll take a look as soon as possible.  In the meantime, if you haven’t already, please carefully read the [issue contribution guidelines](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md#ii-issue-contributions) and double-check for any missing information above.  In particular, please ensure that this issue is about a stability or performance bug with a  documented feature; and make sure you’ve included detailed instructions on how to reproduce the bug from a clean install.   Finally, don’t forget to include the version of Node.js you tested with, as well as your version of Sails or Waterline, and of any relevant standalone adapters/generators/hooks.

Thank you!

**Comment by sgress454** [comment]
_From @slavafomin on August 8, 2016 22:41_

@particlebanana could you please look at it? We've found this issue only when it's reached our production server, and I don't really want to convert our database columns to plain `text` types instead of `enums`. I really hope this issue could be solved promptly.

Thanks!

**Comment by sgress454** [comment]
_From @slavafomin on August 15, 2016 8:24_

Is there anybody still working at this project?

**Comment by sgress454** [comment]
_From @particlebanana on August 15, 2016 15:31_

@slavafomin it's because of the case-sensitivity setting. Try turning that off so the `lower` function doesn't get added on the query.

In your connection config add:

``` javascript
wlNext: {
  caseSensitive: false
}
```

**Comment by sgress454** [comment]
_From @slavafomin on August 16, 2016 14:36_

Hello @particlebanana!

I've changed my config to:

``` js

module.exports.connections = {
  main: {
    adapter  : 'sails-postgresql',
    // ...
    wlNext: {
      caseSensitive: false
    }
  }
};
```

But, I'm still getting the same error.

**Comment by sgress454** [comment]
_From @slavafomin on August 16, 2016 14:45_

I've pinpointed the problem to this module/file: https://github.com/balderdashy/waterline-sequel/blob/master/sequel/lib/criteriaProcessor.js#L382-L384

```
buildParam(): caseSensitive = undefined  <  "user"."email"
buildParam(): caseSensitive = undefined  <  "order"."user"
buildParam(): caseSensitive = true,      <  LOWER("order"."status")
```

Are you sure that sensitivity argument is correctly passed to the `buildParam()` function in all places?

**Comment by sgress454** [comment]
_From @slavafomin on August 21, 2016 23:7_

Any further suggestions on this please?

**Comment by sgress454** [comment]
_From @slavafomin on August 28, 2016 15:43_

UP?

**Comment by sgress454** [comment]
_From @slavafomin on September 2, 2016 0:47_

If someone will get into the same situation as me: the dirty workaround would be to create a `lower` function for each enumeration type. You can do so in PostgreSQL this way:

``` sql
-- enum_users_gender
CREATE OR REPLACE FUNCTION lower(enum_users_gender) RETURNS text AS $$
  SELECT lower($1::text);
$$ LANGUAGE SQL;
```

---

It was a terrible mistake to ever use Sails.js and especially Waterline in our products. Seems like we will have to pay for this dearly...

**Comment by sgress454** [comment]
_From @sailsbot on October 3, 2016 23:0_

@slavafomin,@sailsbot,@particlebanana: Hello, I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message and simply close the issue if possible.  On the other hand, if you are still waiting on a patch, please post a comment to keep the thread alive (with any new information you can provide).

If no further activity occurs on this thread within the next 3 days, the issue will automatically be closed.

Thanks so much for your help!

**Comment by sgress454** [comment]
_From @slavafomin on October 5, 2016 5:1_

@sailsbot keep you hands from it tin can!

**Comment by sgress454** [comment]
_From @sailsbot on November 5, 2016 23:0_

@slavafomin,@sailsbot,@particlebanana: Hello, I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message and simply close the issue if possible.  On the other hand, if you are still waiting on a patch, please post a comment to keep the thread alive (with any new information you can provide).

If no further activity occurs on this thread within the next 3 days, the issue will automatically be closed.

Thanks so much for your help!

**Comment by sgress454** [comment]
_From @slavafomin on November 6, 2016 14:2_

@sailsbot get lost already...

**Comment by sgress454** [comment]
_From @particlebanana on November 7, 2016 18:15_

@slavafomin I'm moving the adapters away from the in house built sql generator to just use Knex. The version on Master should work in this use case - it also removes all the `lower` stuff.

It will be published as `sails-postgresql@1.0.0` when the next version of Waterline is published.

I added the "bug" label to this so the bot will leave it alone.

**Comment by sgress454** [comment]
🎉This was incorporated into the official 1.0 release.
